src/vsr/client_replies.zig

```diff
@@ -129,16 +129,25 @@ pub fn ClientRepliesType(comptime Storage: type) type {
             // Don't unref `write_queue`'s Writes — they are a subset of `writes`.
         }
 
+        pub fn writing_or_queued(client_replies: *ClientReplies, slot: Slot) bool {
+            var writes = client_replies.writes.iterate();
+            while (writes.next()) |write| {
+                if (write.slot.index == slot.index) return true;
+            }
+            assert(!client_replies.writing.isSet(slot.index));
+            return false;
+        }
+
         /// Returns true if the reply at the given slot is  durably persisted to disk. The
         /// difference with `faulty` bit set is that `faulty` is cleared at the start of a write
         /// when the reply is still in RAM. In contrast, `reply_durable` checks that the
         /// corresponding reply hit the disk.
         pub fn reply_durable(
-            client_replies: *const ClientReplies,
+            client_replies: *ClientReplies,
             slot: Slot,
         ) bool {
             return !client_replies.faulty.isSet(slot.index) and
-                !client_replies.writing.isSet(slot.index);
+                !client_replies.writing_or_queued(slot);
```

Comment:
This one I think was a genuine bug which could still trigger storage non-determinism at checkpoint